File: CK/DeOnThi/GD_Momentum.cpp
#include <iostream>
using namespace std;
double e = M_E;
float f(float x)
{
    return pow((exp(2 * x * x + 1) - 2 * x + 1), 2) - 5 * pow(x + 1, 3);
}

float grad(float x)
{
    return 2*(pow(e,4*x) - 2*pow(x,2) - 5*pow(x,3))*(4*pow(e,4*x) - 4*x - 15*pow(x,2)) - 8*pow(x,3);
}
void gradient_momentum(float x, float theta, float alpha, float n, float e)
{
    float x_new, x0 = 0;
    for (int i = 0; i < n; i++)
    {
        float delta = x - x0;
        x_new = x - theta * grad(x) - alpha * delta;
        if (abs(grad(x)) <= e)
        {
            break;
        }
        x0 = x;
        x = x_new;
        cout << "x[" << i + 1 << "] = " << x << endl;
    }
}
int main()
{
    float theta, alpha, e, n;
    float x;
    cout << "Nhap x: " << endl;
    cin >> x;
    cout << "Nhap tham so hoc (learning rate) : " << endl;
    cin >> theta;
    cout << "Nhap he so dong luong : " << endl;
    cin >> alpha;
    cout << "Nhap so buoc lap : " << endl;
    cin >> n;
    cout << "Nhap sai so : " << endl;
    cin >> e;
    gradient_momentum(x, theta, alpha, n, e);
}

File: CK/DeOnThi/GD_Momentum.py
from __future__ import division, print_function, unicode_literals
import math
import numpy as np

# f(x)


def cost(x):
    return (math.exp(4*x) - 2*(x**2) - 5*(x**3))**2 + 2*(x**4)


# f'(x)
def grad(x):
    return 2*(math.exp(4*x) - 2*(x**2) - 5*(x**3))*(4*math.exp(4*x) - 4*x - 15*(x**2)) + 8*(x**3)


def GD_momentum(theta_init, grad, learningRate, momentum, e, loop):
    theta = [theta_init]
    v_old = np.zeros_like(theta_init)
    for it in range(loop):
        v_new = momentum*v_old + learningRate*grad(theta[-1])
        theta_new = theta[-1] - v_new
        if abs(grad(theta_new)) < e:
            break
        theta.append(theta_new)
        v_old = v_new
    return theta, it


if __name__ == '__main__':
    x0 = input("Nhap diem bat dau x0: ")
    x0 = float(x0)
    learningRate = input("Nhap tham so hoc learning rate: ")
    learningRate = float(learningRate)
    momentum = input("Nhap tham so momentum: ")
    momentum = float(momentum)
    e = input("Nhap do chinh xac e: ")
    e = float(e)
    loop = input("Nhap so lan lap loop: ")
    loop = int(loop)
    (x, it) = GD_momentum(x0, grad, learningRate, momentum, e, loop)

    print('Gia tri nho nhat tai x = %f, min(fx) = %f' % (x[-1], cost(x[-1])))
